### PR TITLE
Issue2

### PR DESCRIPTION
**Pull request by marouaneamz**
#2

### CODE REVIEW

**Comment by marouaneamz** [comment]
ERROR: test_loss (tests.test_models.test_heads.TestMultiTaskHead)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/marouane/dev/mmclassification1.x/tests/test_models/test_heads.py", line 518, in test_loss
    losses = head.loss(feats, data_samples)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/heads/multi_task_head.py", line 83, in loss
    masked_features = masked_features + (feature[mask], )
IndexError: The shape of the mask [1] at index 0 does not match the shape of the indexed tensor [4, 10] at index 0

**Comment by piercus** [comment]
@marouaneamz I see code related the metrics in this PR, can you please create 2 different issues, and keep only multi_task_head related code in here

**Comment by marouaneamz** [comment]
> @marouaneamz I see code related the metrics in this PR, can you please create 2 different issues, and keep only multi_task_head related code in here

done

File: tests/test_models/test_heads.py [hunk 1]
```diff
@@ -480,3 +480,41 @@ def test_forward(self):
         # return the last item (same as pre_logits)
         feats = (torch.rand(4, 10), torch.rand(4, 10))
         head(feats)
+
+class TestMultiTaskHead(TestCase):
+    DEFAULT_ARGS = dict(
+        type='MultiTaskHead',                                    # <- Head config, depends on #675
+        sub_heads={
```
Comment (piercus) [inline tests/test_models/test_heads.py]:
```suggestion
        task_heads={
```

File: tests/test_models/test_heads.py [hunk 2]
```diff
@@ -480,3 +480,41 @@ def test_forward(self):
...
+        self.assertTrue(isinstance(outs, tuple))
+
+
+    def test_loss(self):
+        feats = (torch.rand(4, 10),)
+        gt_label = {
+            'task1': torch.randint(0, 3, (4,)),
+            'task2': torch.randint(0, 6, (4,)),
+        }
+        data_samples = [MultiTaskDataSample(gt_label.keys()).set_gt_label(gt_label)]
```
Comment (piercus) [inline tests/test_models/test_heads.py]:
```suggestion
        data_samples = []
        
        for _ in range(self.BATCH_SIZE):
          gt_label = {}
          for task_name, _ in self.DEFAULT_ARGS.task_heads:
            gt_label[task_name] = 1
          data_sample = MultiTaskDataSample().set_gt_label(gt_label)
          data_samples.append(data_sample)
```

**Review by piercus — CHANGES_REQUESTED** [review]
Hello @marouaneamz 

I feel this is a good start, check my inline comments.

## More unit tests

### `test_loss_empty_data_sample`
When data_sample is empty then the loss should be 0 (we will consider that the values of the heads are masked) 

### `test_loss_invalid_data_sample_task_name`

If data_sample has keys which are not in the heads, then it should throw an error `The data_sample task "task2" is not matching with the heads task names (task0, task1)`

### `test_nested_multi_task_loss`

If you use a MultiTaskHead in a MultiTaskHead.

With a structure like 

* multiTaskHead
  * `task0` : multiTaskHead
    * `task00` : LinearClsHead
    * `task01` : LinearClsHead
  * `task1` : LinearClsHead

 then it should work with 

```python
gt_label = {
  'task0' : {
    'task00': 1,
    'task01': 1
  },
  'task1': 1
}
```
### `test_nested_invalid_sample1`

Same head than `test_nested_multi_task_loss` but with a gt_label like

```python
gt_label = {
  'task0' : 1,
  'task1': 1
}
```

It should fail saying `values must be a dictionnary (currently is "1")` in MultiTaskDataSample.set_gt_label

### `test_nested_invalid_sample2`

Same head than `test_loss` but with a gt_label like

```python
gt_label = {
  'task0' : {
    'task00': 1,
    'task01': 1
  },
  'task1': 1
}
```

It should fail saying `values must be a a number (currently is "{'task00': 1,'task01': 1}")` in MultiTaskDataSample.set_gt_label

## MultiTaskDataSample

Considering MultiTaskDataSample is used in the unit test, please add unit test for MultiTaskDataSample in this PR also

File: tests/test_models/test_heads.py [hunk 3]
```diff
@@ -480,3 +480,148 @@ def test_forward(self):
...
+        task_heads={
+            'task0': dict(type='MultiTaskHead',
+                task_heads = {
+                    'task00': dict(type='LinearClsHead', num_classes=3),
+                    'task01': dict(type='LinearClsHead', num_classes=6),
+                    },
+                common_cfg=dict(
+                    in_channels=10,
+                    loss=dict(type='CrossEntropyLoss', loss_weight=1.0)
+                    ),
```
Comment (piercus) [inline tests/test_models/test_heads.py]: please move common_cfg to the upper dict

File: mmcls/models/heads/multi_task_head.py [hunk 4]
```diff
@@ -9,35 +9,46 @@
...
-            self.sub_heads[task_name] = sub_head
+        for task_name, head_cfg in task_heads.items():
+            if head_cfg['type'] != 'MultiTaskHead':
+                sub_head = MODELS.build(head_cfg, default_args=common_cfg)
+            else:
+                sub_head = MODELS.build(
+                    head_cfg, default_args={'common_cfg': common_cfg})
+            self.task_heads[task_name] = sub_head
 
     def forward(self, feats):
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py:90]: Please adda documentation here explaining input and output format

File: mmcls/models/heads/multi_task_head.py [hunk 5]
```diff
@@ -59,22 +70,14 @@ def loss(self, feats: Tuple[torch.Tensor],
             dict[str, Tensor]: a dictionary of loss components
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py]: Please update the loss documentation

> The features extracted from the backbone.

Ok

> Multiple stage inputs are acceptable but only the last stage
> will be used to classify. The shape of every item should be
> ``(num_samples, num_classes)``.

Not true anymore


```suggestion
            dict[str, Tensor]: a dictionary of loss components, each task loss key will be prefixed by the task_name like "task1_loss"
```

File: mmcls/models/heads/multi_task_head.py [hunk 6]
```diff
@@ -83,9 +86,11 @@ def loss(self, feats: Tuple[torch.Tensor],
...
             else:
                 masked_features = feats[mask]
-            head_loss = head.loss(masked_features, masked_data_samples,
-                                  **kwargs)
-
+            if len(masked_data_samples) > 0:
+                head_loss = head.loss(masked_features, masked_data_samples,
+                                      **kwargs)
+            else:
+                head_loss = {'loss': torch.tensor(0)}
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py]: How do we know that the name of the loss is "loss" ?

```suggestion
                head_loss = {'loss': torch.tensor(0)}
```

File: mmcls/models/heads/multi_task_head.py [hunk 7]
```diff
@@ -9,35 +9,46 @@
...
-        for task_name, head_cfg in sub_heads.items():
-            sub_head = MODELS.build(head_cfg, default_args=common_cfg)
-            self.sub_heads[task_name] = sub_head
+        for task_name, head_cfg in task_heads.items():
+            if head_cfg['type'] != 'MultiTaskHead':
+                sub_head = MODELS.build(head_cfg, default_args=common_cfg)
+            else:
+                sub_head = MODELS.build(
+                    head_cfg, default_args={'common_cfg': common_cfg})
+            self.task_heads[task_name] = sub_head
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py:88]: Please apply the mmtask_convertor directly to the head here as suggested on slack by haochen ye on openmmlab slack

File: mmcls/models/heads/multi_task_head.py [hunk 8]
```diff
@@ -59,22 +70,14 @@ def loss(self, feats: Tuple[torch.Tensor],
...
-              mask = to_tensor([True]*batch_n)
-            """
+        for task_name, head in self.task_heads.items():
             mask = []
             masked_data_samples = []
             for data_sample in data_samples:
                 sample_mask = data_sample.get_task_mask(task_name)
                 if sample_mask:
-                    sample = data_sample.get_task_sample(task_name)
+                    sample = mmtask_convertor(head.loss, task_name,
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py]: I prefer we do not change the sample here, but directly wrap the head.loss as discussed on slack

**Review by piercus — CHANGES_REQUESTED** [review]
see comments

File: mmcls/models/heads/multi_task_head.py (hunk 6 above)
Comment (marouaneamz) [reply mmcls/models/heads/multi_task_head.py]: https://github.com/marouaneamz/mmclassification/blob/31c67ffed406331ab232796b3fdb72f0684b1442/mmcls/models/heads/cls_head.py#L94

Comment (marouaneamz) [reply mmcls/models/heads/multi_task_head.py]: Normally it should be  head_loss = {task_name_'loss': torch.tensor(0)}

File: mmcls/models/heads/multi_task_head.py [hunk 9]
```diff
@@ -9,35 +9,88 @@
...
+                        data_sample.to_target_data_sample(
+                            target_type, task_name))
+        masked_inputs = tuple()
+        if type(inputs) is tuple:
+            for input in inputs:
+                masked_inputs = masked_inputs + (input[mask], )
+        else:
+            masked_inputs = inputs[mask]
+        if len(task_data_samples) == 0:
+            return {'loss': torch.tensor(0), 'mask_size': 'all'}
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py]:
```suggestion
            return {'loss': torch.tensor(0), 'mask_size': 0}
```

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_models/test_heads.py [hunk 10]
```diff
@@ -480,3 +480,139 @@ def test_forward(self):
...
+                gt_label[task_name] = 1
+            data_sample = MultiTaskDataSample().set_gt_task(gt_label)
+            data_samples.append(data_sample)
+        head = MODELS.build(self.DEFAULT_ARGS)
+        # with without data_samples
+        predictions = head.predict(feats)
+        self.assertTrue(is_seq_of(predictions, MultiTaskDataSample))
+        for pred in predictions:
+            self.assertIn('task0', pred.pred_task)
+
```
Comment (piercus) [inline tests/test_models/test_heads.py:564]: can you please add a test like 

```
task0_sample = predictions[0].to_target_data_sample('ClsDataSample', 'task0')
self.assertTrue(isinstance(task0_sample.pred_label, torch.tensor))
```

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_models/test_heads.py (hunk 10 above)
Comment (marouaneamz) [reply tests/test_models/test_heads.py:564]: I think it should be like this
```
task0_sample = predictions[0].to_target_data_sample('ClsDataSample', 'task0')
self.assertTrue(isinstance(task0_sample.pred_label.score, torch.tensor))
```

File: mmcls/structures/multi_task_data_sample.py [hunk 11]
```diff
@@ -39,7 +39,14 @@ def set_gt_task(self, value: Dict) -> 'MultiTaskDataSample':
 
     def set_pred_task(self, value: Dict) -> 'MultiTaskDataSample':
         """Set score of ``pred_task``."""
-        self.pred_task = LabelData(**value)
+        if list(value.values())[0] == torch.tensor:
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: maybe you should check the type, if not seen it means that it is not covered by any unit test, please cover this and fix this

Comment (piercus) [reply mmcls/structures/multi_task_data_sample.py]: What happens in value is an empty dict ? please add a unit test, it should not fail, set self.pred_task to `LabelData`

File: mmcls/structures/multi_task_data_sample.py [hunk 12]
```diff
@@ -39,7 +39,14 @@ def set_gt_task(self, value: Dict) -> 'MultiTaskDataSample':
...
+            self.pred_task = LabelData(**value)
+        else:
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: I would suggest a more strict approah here like 
```suggestion
        elif type(list(value.values())[0]) in self.data_samples_map.keys():
```

And raise an exception if the type is not known

File: mmcls/structures/multi_task_data_sample.py [hunk 13]
```diff
@@ -97,3 +104,13 @@ def to_target_data_sample(self, target_type, task_name):
...
     }
+
+    def from_target_data_sample(self, target_type, value):
+        return self.data_samples_map2[target_type](self, value)
+
+    def from_cls_data_sample(self, value):
+        return value.pred_label.score
+
+    data_samples_map2 = {
+        'ClsDataSample': from_cls_data_sample,
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: Why is there no implementation `from_multi_task_data_sample` ?

File: mmcls/structures/multi_task_data_sample.py [hunk 14]
```diff
@@ -97,3 +104,13 @@ def to_target_data_sample(self, target_type, task_name):
         'ClsDataSample': to_cls_data_sample,
         'MultiTaskDataSample': to_multi_task_data_sample
     }
+
+    def from_target_data_sample(self, target_type, value):
+        return self.data_samples_map2[target_type](self, value)
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: I suggest you put everything in self.data_samples_map

like 
```
data_samples_map = {
￼        'ClsDataSample': {to: to_cls_data_sample, from: from_cls_data_sample},		￼        
         'MultiTaskDataSample': {to: to_multi_task_data_sample, from: from_multi_task_data_sample},	
￼    }
```

**Review by piercus — CHANGES_REQUESTED** [review]

**Review by piercus — APPROVED** [review]

**Pull request merged**